File: python/sum_lists.py
from linked_list import LinkedList
from node import Node

"""
You have two numbers represented by a linked list, where each node contains a single
digit. The digits are stored in reverse order, such that the 1st digit is at the head.
Write a function that adds the two numbers and returns the sum as a linked list.

FOLLOW UP
Suppose the digits are stored in forward order. Repeat the above problem.
"""

def sum_lists_using_heads(node_longest_list, node_shortest_list, value_carried):
    while node_longest_list:
        value_1 = node_longest_list.data
        value_2 = node_shortest_list.data if node_shortest_list else 0
        result = value_1 + value_2 + value_carried
        value_carried = 0

        if result > 9:
            value_carried = int(str(result)[0])
            result = int(str(result)[1])

        node_longest_list.data = result
        node_longest_list = node_longest_list.next
        node_shortest_list = node_shortest_list.next if node_shortest_list else None


# time complexity: O(n), space complexity: O(1)
def sum_lists(list_1, list_2):
    node_longest_list = list_1.head if list_1.count > list_2.count else list_2.head
    node_shortest_list = list_2.head if list_2.count < list_1.count else list_1.head
    value_carried = 0

    sum_lists_using_heads(node_longest_list, node_shortest_list, value_carried)

    if value_carried:
        new_node = Node(value_carried)
        node_longest_list.next = new_node

    return list_1 if list_1.count > list_2.count else list_2


def sum_lists_forward_using_heads(longest_list, node_longest_list, node_shortest_list, value_carried):
    prev_node_longest_list = None
    while node_longest_list:
        value_1 = node_longest_list.data
        value_2 = node_shortest_list.data if node_shortest_list else 0
        result = value_1 + value_2

        if result > 9:
            value_carried = int(str(result)[0])
            result = int(str(result)[1])
            if prev_node_longest_list:
                prev_node_longest_list.data += value_carried
            else:
                longest_list.add(value_carried)

        node_longest_list.data = result
        prev_node_longest_list = node_longest_list
        node_longest_list = node_longest_list.next
        node_shortest_list = node_shortest_list.next if node_shortest_list else None


# time complexity: O(n), space complexity: O(1)
def sum_lists_forward(list_1, list_2):
    longest_list = list_1 if list_1.count > list_2.count else list_2
    shortest_list = list_2 if list_2.count < list_1.count else list_1
    value_carried = 0
    num_paddings = abs(list_1.count - list_2.count)

    for i in range(num_paddings):
        shortest_list.add(0)

    sum_lists_forward_using_heads(longest_list, longest_list.head, shortest_list.head, value_carried)
    return longest_list


list = LinkedList(6)
list.add(1)
list.add(7)

list_2 = LinkedList(2)
list_2.add(9)
list_2.add(5)

result_list = sum_lists(list, list_2)
result_list.print_data()


list_forward = LinkedList(7)
list_forward.add(1)
list_forward.add(6)

list_forward_2 = LinkedList(5)
list_forward_2.add(9)
list_forward_2.add(2)

result_list = sum_lists_forward(list_forward, list_forward_2)
result_list.print_data()

File: java/SumLists.java
import java.lang.Math;

class SumLists {
    public static void main(String args[]) {
        MyLinkedList<Integer> list = new MyLinkedList<Integer>(6);
        list.add(1);
        list.add(7);

        MyLinkedList<Integer> list2 = new MyLinkedList<Integer>(2);
        list2.add(9);
        list2.add(5);

        MyLinkedList<Integer> resultList = sumLists(list, list2);
        resultList.printData();

        MyLinkedList<Integer> listForward = new MyLinkedList<Integer>(7);
        listForward.add(1);
        listForward.add(6);

        MyLinkedList<Integer> listForward2 = new MyLinkedList<Integer>(5);
        listForward2.add(9);
        listForward2.add(2);

        MyLinkedList<Integer> resultListForward = sumListsForward(listForward, listForward2);
        resultListForward.printData();
    }

    private static void sumListsUsingHeads(
            MyNode<Integer> nodeLongestList,
            MyNode<Integer> nodeShortestList,
            int valueCarried)
    {
        while (nodeLongestList != null) {
            int value1 = nodeLongestList.data;
            int value2 = nodeShortestList != null ? nodeShortestList.data : 0;
            int result = value1 + value2 + valueCarried;
            valueCarried = 0;

            if (result > 9) {
                valueCarried = Integer.parseInt(Integer.toString(result).substring(0, 1));
                result = Integer.parseInt(Integer.toString(result).substring(1, 2));
            }
            nodeLongestList.data = result;
            nodeLongestList = nodeLongestList.next;
            nodeShortestList = nodeShortestList != null ? nodeShortestList.next : null;
        }
    }

    private static MyLinkedList sumLists(MyLinkedList list1, MyLinkedList list2) {
        MyNode<Integer> nodeLongestList = list1.length > list2.length ? list1.head : list2.head;
        MyNode<Integer> nodeShortestList = list2.length < list1.length ? list2.head : list1.head;
        int valueCarried = 0;

        sumListsUsingHeads(nodeLongestList, nodeShortestList, valueCarried);

        if (valueCarried > 0) {
            MyNode<Integer> newNode = new MyNode(valueCarried);
            nodeLongestList.next = newNode;
        }
        return list1.length > list2.length ? list1 : list2;
    }

    private static void sumListsForwardUsingHeads(
            MyLinkedList longestList,
            MyNode<Integer> nodeLongestList,
            MyNode<Integer> nodeShortestList,
            int valueCarried)
    {
        MyNode<Integer> prevNodeLongestList = null;
        while (nodeLongestList != null) {
            int value1 = nodeLongestList.data;
            int value2 = nodeShortestList != null ? nodeShortestList.data : 0;
            int result = value1 + value2;

            if (result > 9) {
                valueCarried = Integer.parseInt(Integer.toString(result).substring(0, 1));
                result = Integer.parseInt(Integer.toString(result).substring(1, 2));
                if (prevNodeLongestList != null) {
                    prevNodeLongestList.data += valueCarried;
                } else {
                    longestList.add(valueCarried);
                }
            }
            nodeLongestList.data = result;
            prevNodeLongestList = nodeLongestList;
            nodeLongestList = nodeLongestList.next;
            nodeShortestList = nodeShortestList != null ? nodeShortestList.next : null;
        }
    }

    private static MyLinkedList sumListsForward(MyLinkedList list1, MyLinkedList list2) {
        MyLinkedList longestList = list1.length > list2.length ? list1 : list2;
        MyLinkedList shortestList = list2.length < list1.length ? list2 : list1;
        int valueCarried = 0;
        int numPaddings = Math.abs(list1.length - list2.length);

        for (int i = 0; i < numPaddings; i++) {
            shortestList.add(0);
        }

        sumListsForwardUsingHeads(longestList, longestList.head, shortestList.head, valueCarried);
        return longestList;
    }
}

